crates/turbopack-core/src/resolve/mod.rs

```diff
@@ -2188,6 +2188,59 @@ async fn resolve_module_request(
 
     let mut results = vec![];
 
+    // Self references, if the nearest package.json has the name of the requested
+    // module. This matches only using the exports field and no other
+    // fields/fallbacks.
+    let package_json_context = find_context_file(lookup_path, package_json()).await?;
+    if let FindContextFileResult::Found(package_json_path, _refs) = &*package_json_context {
+        let read = read_package_json(*package_json_path).await?;
+        if let Some(json) = &*read {
+            if json["name"].as_str() == Some(module) {
```

Comment:
To improve caching on that part, I think it's worth moving this part into a separate turbo-tasks function that returns the potential self reference for a lookup_path:

``` rs
#[turbo_tasks::value]
enum SelfReferencePackageResult {
    Found {
        name: String,
        path: Vc<FileSystemPath>,
    }
    NotFound
}

fn find_self_reference(lookup_path: Vc<FileSystemPath>) -> Vc<SelfReferencePackageResult>
```